proto/backend.proto

```diff
@@ -155,6 +155,9 @@ message DataResponse {
   // The status codes should match values from standard HTTP status codes
   // If not set explicitly, it will be marshaled to 200 if no error exists, or 500 if one does
   int32 status = 4;
+
+  // Error source
+  string source = 5;
```

Comment:
It just struck me that we need to discuss this in relation to the [plugin request duration metric](https://github.com/grafana/grafana/blob/13db9066a39db4a92fe5090f9b9336900f7d432b/pkg/plugins/backendplugin/instrumentation/instrumentation.go#L41-L46). Your changes only covers query data and that's the first drawback. Second, in theory there could be multiple data responses with errors and different sources while the plugin metrics only covers one request/one response - how are you imagine this to work?

For a returned error from check health, call resource and query data request the error source needs to somehow be embedded within the error for it to be able to cross the gRPC boundary and for Grafana to recognize the error source. Any idea how to achieve this?

Reply:
@bergquist can you chime in here?

Reply:
After a quick discussion in the plugins working group, there might be a chance that with multiple errors, it could report a Grafana error as an external one, which would be damaging. Could we make an easy guess at how often we might expect multiple data response errors and different data sources? (my gut says it's a high percentage).

In general, I think if we have a solution that can improve things even in the case of 1 query data error, we can add in more later. Where we want to be careful is to not give wrong information confidently - we'd want to err on the side of taking a closer look before declaring something if we aren't certain.

Curious to hear @bergquist thoughts as well!

Reply:
I think the likelihood of having both downstream and plugin related errors in the same multi query request is very low.  For example, I issue 3 queries on one request, and one of them fails to connect to the database.   I highly doubt another one will also have an internal plugin error.  Now, we could also prioritize an internal plugin error over external.  So if 2 queries have a "plugin" error, and the third has a "downstream" error, we return "plugin" as the source since something in that request caused a failure in our code.

Reply:
I also think the majority of errors will come from queries.  Yes resource calls and health checks can and will produce errors, but will be less frequent, since most users spend more time viewing dashboards.  So can we address those in separate prs?

Reply:
> I think the likelihood of having both downstream and plugin related errors in the same multi query request is very low.  For example, I issue 3 queries on one request, and one of them fails to connect to the database.   I highly doubt another one will also have an internal plugin error.  Now, we could also prioritize an internal plugin error over external.  So if 2 queries have a "plugin" error, and the third has a "downstream" error, we return "plugin" as the source since something in that request caused a failure in our code.

Priority make sense, good idea.

Reply:
> I also think the majority of errors will come from queries.  Yes resource calls and health checks can and will produce errors, but will be less frequent, since most users spend more time viewing dashboards.  So can we address those in separate prs?

Sure if you say so. Would still recommend you open a PR in grafana repository that uses this branch as sdk dependency and implement needed changes there. This will simplify I think. Or are you're goal starting to use this in your plugins before it's being used in Grafana?

Reply:
@marefr you ok with addressing health check and resource calls separately?

Reply:
> @marefr you ok with addressing health check and resource calls separately?

Sure. But would be good if we have an idea how to support this for those endpoints so maybe starting a small doc would be good perhaps to continue the discussions without blocking this

Reply:
> Or are you're goal starting to use this in your plugins before it's being used in Grafana?

Yes we want to start using this in external plugins asap.

Reply:
> Now, we could also prioritize an internal plugin error over external. So if 2 queries have a "plugin" error, and the third has a "downstream" error, we return "plugin" as the source since something in that request caused a failure in our code.

This makes sense to me. 

The contract between grafana and the plugin is one grpc call regardless of how many queries and right now I think that is what we should track. There has been discussions around tracking per query but I don't think this would block that in the future.